### PR TITLE
Encrypted URL cache

### PR DESCRIPTION
**Pull request by bbirman**

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.h [hunk 1]
```diff
@@ -319,6 +319,16 @@ NS_SWIFT_NAME(SalesforceManager)
  */
 - (NSArray<NSString *>*)getDevSupportInfos NS_SWIFT_NAME(devSupportInfoList());
 
+/**
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.h]: Maybe we could just have a property?
```objective-c
@property (nonatomic, assign) BOOL encryptURLCache;
```

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFEncryptionKey.h [hunk 2]
```diff
@@ -49,14 +49,14 @@ NS_ASSUME_NONNULL_BEGIN
  @param dataToEncrypt The data to encrypt.
  @returns `NSData` object that contains the encrypted data.
  */
-- (NSData*)encryptData:(NSData *)dataToEncrypt;
+- (nullable NSData *)encryptData:(NSData *)dataToEncrypt;
```
Comment (bbirman) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFEncryptionKey.h]: The methods that these call into are nullable but if it's too much of a semver issue I can leave them

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.m [hunk 3]
```diff
@@ -207,10 +213,27 @@ - (instancetype)init {
...
         [self setupServiceConfiguration];
+        [self enableEncryptedURLCache];
     }
     return self;
 }
 
+- (void)enableEncryptedURLCache {
+    if (![NSURLCache.sharedURLCache isKindOfClass:[SFSDKEncryptedURLCache class]]) {
+        [NSURLCache.sharedURLCache removeAllCachedResponses];
+        SFSDKEncryptedURLCache *encryptedCache = [[SFSDKEncryptedURLCache alloc] initWithMemoryCapacity:kDefaultCacheMemoryCapacity diskCapacity:kDefaultCacheDiskCapacity diskPath:kDefaultCachePath];
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.m]: Should we provide a way for the app to pass those values (memory and disk capacity)?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    NSURLRequest *requestWithSecureURL = [self requestWithSecureURLForRequest:request];
+    if (!requestWithSecureURL) {
+        [SFSDKCoreLogger d:[self class] format:@"RequestWithSecureURL is nil"];
+        return;
+    }
+    
+    // For cachedResponse.data
+    NSData *encryptedResponseData = [self.encryptionKey encryptData:cachedResponse.data];
+    if (!encryptedResponseData) {
+        [SFSDKCoreLogger d:[self class] format:@"Unable to encrypt response"];
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m]: Should log level be d(ebug) or e(rror)?

File: libs/SalesforceSDKCore/SalesforceSDKCore/SFSDKEncryptedURLCacheTests.m [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    Method useCache_networkForRequest = class_getInstanceMethod([self class], @selector(useCache_networkForRequest:));
+    
+    // Calls that will hit network and should store into cache for next call to use
+    method_exchangeImplementations(networkForRequest, ignoreCache_networkForRequest);
+    [self makeRequests];
+    
+    // Calls that will only load from cache
+    method_exchangeImplementations(networkForRequest, useCache_networkForRequest);
+    [self makeRequests];
+}
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/SFSDKEncryptedURLCacheTests.m]: Shouldn't you put the original method back at the end of the test?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.m (hunk 3 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.m]: The app can change it through sharedURLCache (`NSURLCache.sharedURLCache.memoryCapacity`) but let me know if that's not good enough

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m (hunk 4 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m]: 👍I'll change to error. Also, what do you think about trying to remove the item from the cache if decrypting fails?

Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m]: You mean in `cachedResponseForRequest`? If the cache returns nil, wouldn't the request be sent to the server and the server's response be written in the cache (over the one we couldn't decrypt)?

**Comment by bbirman** [comment]
Looking into test failures

**Pull request merged**